protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+- When making changes to a contract, always bump to the lowest possible version based on the specific change you are making. We do not want to e.g. optimistically bump to a major version, because protocol sequencing may change unexpectedly. Use these examples to know how to bump the version:
+  - Example 1: A contract is currently on `1.2.3`. We don't yet know when the next release of this contract will be. However, you are simply fixing typos in comments so you bump the version to `1.2.4-beta.1`.
+    - The next PR that makes in that same contract but makes no functional changes, so it bumps the version to `1.2.4-beta.2`.
+    - The last PR introduces a breaking change, which bumps the version from `1.2.4-beta.2` to `2.0.0-beta.1`.
+  - Example 2: A contract is currently on `2.4.7`. We know the next release of this contract will be a breaking change. Regardless, as you start development by fixing typos in comments, bump the version to `2.4.8-beta.1`. This is because we may end up putting out a release before the breaking change is added.
+    - Once you start working on the breaking change, bump the version to `3.0.0-beta.1`.
+    - Continue to bump the beta version as you make changes. When the contract is ready for release, bump the version to `3.0.0-rc.1`.
+- New contracts start at `0.y.z` and only become `1.0.0` when they are ready for production.
+
+Versioning for monorepo releases:
+
+- Monorepo releases continue to follow the `op-contracts/vX.Y.Z` naming convention.
+- The version used for the next release is determined by the highest version bump of any individual contract in the release.
+  - Example 1: The monorepo is at `op-contracts/v1.5.0`. Various tech debt is cleaned up in contracts, but no features are added, so all contracts only bump the patch version. The next monorepo release will be `op-contracts/v1.5.1`.
+  - Example 2: The monorepo is at `op-contracts/v1.5.1`. Legacy `ALL_CAPS()` getter methods are removed from a contract, causing that contract to bump the major version. The next monorepo release will be `op-contracts/v2.0.0`.
+- A monorepo contracts release must map to an exact set of contract semvers, and this mapping must be defined in the contract release notes which are the source of truth. See [`op-contracts/v1.4.0-rc.4`](https://github.com/ethereum-optimism/optimism/releases/tag/op-contracts%2Fv1.4.0-rc.4) for an example of what release notes should look like.
+
+The above gets us most of the way there, and there is one problem not yet addressed: When a release is proposed to governance, we propose a commit hash, and often those proposed contracts are already deployed to mainnet.
+For example, the [Fault Proofs governance proposal](https://gov.optimism.io/t/upgrade-proposal-fault-proofs/8161) provides specific addresses that will be used.
+This is ideal and we should continue to do that, but we need to fit this in to the above scheme.
+
+To accommodate this, once ready for governance approval the release flow is:
```

Comment:
To me this is the weakest part of the proposal and I would to get some thoughts on feedback on how to handle this

Reply:
This may actually be a strength. It provides a dedicated branch for each release, making it easy to deploy a patch or bug fix, regardless of other changes that may have occurred on `develop` since the release.